File: static/code/sample/tethered/gdl-cmethernet-c.cs
// In this sample:
// Initialize the network profile (IP, gateway, etc.).
// Implement Listen, Accept, Send, Receive, and Available.
// Send the message: "Hello, I am DUELink\r\n" to the PC.
// Receive the message from the PC: "Hello, I am PC".
// At the end of the sample, C# code is provided to set up the PC.

using System.Net;
using System.Net.Sockets;
using System.Text;
using GHIElectronics.DUELink;

var availablePort = DUELinkController.GetConnectionPort();
var duelink = new DUELinkController(availablePort);

int Initialize(string ip_add, string gw_add) {
    var ip = IPAddress.Parse(ip_add).GetAddressBytes();
    var gw = IPAddress.Parse(gw_add).GetAddressBytes();

    var cmd = $"Init([{ip[0]},{ip[1]},{ip[2]},{ip[3]}],[{gw[0]},{gw[1]},{gw[2]},{gw[3]}],[255,255,255,0])";
    var ret = duelink.Engine.ExecuteCommand(cmd);
    return (int)ret;
}

int Listen(int socket, int port) {
    var cmd = $"Listen({socket},{port})";
    var ret = duelink.Engine.ExecuteCommand(cmd);
    return (int)ret;
}

int Accept(int socket) {
    var cmd = $"Accept({socket})";
    var ret = duelink.Engine.ExecuteCommand(cmd);
    return (int)ret;
}

int Send(int socket, string data) {
    var to_write = UTF8Encoding.UTF8.GetBytes(data);
    duelink.Engine.ExecuteCommand($"dim b6[{to_write.Length}]");

    for (var i = 0; i < to_write.Length; i++ ) {
        duelink.Engine.ExecuteCommand($"b6[{i}]={to_write[i]}");
    }
    
    var cmd = $"Send({socket},b6)";
    var ret = duelink.Engine.ExecuteCommand(cmd);
    return (int)ret;
}

int Receive(int socket, byte[] data) {
    duelink.Engine.ExecuteCommand($"dim b7[{data.Length}]");
    var rev_len = (int)duelink.Engine.ExecuteCommand($"Recv({socket},b7)");
    
    var len = rev_len < data.Length ? rev_len : data.Length;

    for (var i = 0; i < len; i++) {
        data[i] = (byte)duelink.Engine.ExecuteCommand($"b7[{i}]");
    }

    return len;
}

int Available(int socket) {
    var ret = (int)duelink.Engine.ExecuteCommand($"Avail({socket})");
    return ret;
}

Initialize("192.168.0.50", "192.168.0.1");
var socket = 0;
var port = 5000;
while (true) {
    Thread.Sleep(1000);

    Listen(socket, port);
    Console.WriteLine("Listening...");
    Accept(socket);

    Console.WriteLine("Socket connected!");
    Send(socket, "Hello, I am DUELink\r\n");

    while (true) {
        Thread.Sleep(1);
        var read = Available(socket);

        if (read != 0) {
            var data = new byte[read];
            Receive(socket, data );
            var msg = Encoding.UTF8.GetString(data);
            if (msg != null) {
                Console.WriteLine(msg);
            }           
        }
    }
}

//This is for setup the PC, using C#
// 
//using System.Net.Sockets;
//using System.Text;

//var serverIp = "192.168.0.50"; // Replace with your server IP
//var port = 5000; // Replace with your server port
//try {
//    using (var client = new TcpClient(serverIp, port))
//    using (var stream = client.GetStream()) {
//        // Send "Hello World"
//        var message = "Hello, I am PC";
//        var dataToSend = Encoding.UTF8.GetBytes(message);
//        stream.Write(dataToSend, 0, dataToSend.Length);
//        Console.WriteLine("Sent: " + message);

//        // Receive response
//        var receivedBuffer = new byte[1024];
//        var bytesRead = stream.Read(receivedBuffer, 0, receivedBuffer.Length);
//        var response = Encoding.UTF8.GetString(receivedBuffer, 0, bytesRead);
//        Console.WriteLine("Received: " + response);
//    }
//}
//catch (Exception ex) {
//    Console.WriteLine("Error: " + ex.Message);
//}

File: static/code/sample/tethered/gdl-cmethernet-c.js
// In this sample:
// Implement Listen, Accept, Send, Receive, and Available.
// Send the message: "Hello, I am DUELink\r\n" to the PC.
// Receive the message from the PC: "Hello, I am PC".
// Check https://www.duelink.com/docs/products/cmethernet-c?show=tie-samples&lang=dotnet for reference on setting up the client on a PC.

import pkg_serialusb from 'dlserialusb';
const {SerialUSB} = pkg_serialusb

import pkg_duelink from 'duelink';
const {DUELinkController} = pkg_duelink

let duelink = new DUELinkController(new SerialUSB());
await duelink.Connect();

// Sleep helper (ms)
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Methods
async function Initialize(ip_add, gw_add) {
    const ip = ip_add.split('.').map(x => parseInt(x));
    const gw = gw_add.split('.').map(x => parseInt(x));

    const cmd = `Init([${ip[0]},${ip[1]},${ip[2]},${ip[3]}],[${gw[0]},${gw[1]},${gw[2]},${gw[3]}],[255,255,255,0])`;
    const ret = await duelink.Engine.ExecuteCommand(cmd);
    return parseInt(ret);
}

async function Listen(socket, port) {
    const cmd = `Listen(${socket},${port})`;
    const ret = await duelink.Engine.ExecuteCommand(cmd);
    return parseInt(ret);
}

async function Accept(socket) {
    const cmd = `Accept(${socket})`;
    const ret = await duelink.Engine.ExecuteCommand(cmd);
    return parseInt(ret);
}

async function Send(socket, data) {
    const to_write = Buffer.from(data, 'utf-8');
    await duelink.Engine.ExecuteCommand(`dim b6[${to_write.length}]`);

    for (let i = 0; i < to_write.length; i++) {
        await duelink.Engine.ExecuteCommand(`b6[${i}]=${to_write[i]}`);
    }

    const cmd = `Send(${socket},b6)`;
    const ret = await duelink.Engine.ExecuteCommand(cmd);
    return parseInt(ret);
}

async function Receive(socket, data) {
    await duelink.Engine.ExecuteCommand(`dim b7[${data.length}]`);
    const rev_len = parseInt(await duelink.Engine.ExecuteCommand(`Recv(${socket},b7)`));

    const len = rev_len < data.length ? rev_len : data.length;

    for (let i = 0; i < len; i++) {
        data[i] = parseInt(await duelink.Engine.ExecuteCommand(`b7[${i}]`));
    }

    return len;
}

async function Available(socket) {
    const ret = parseInt(await duelink.Engine.ExecuteCommand(`Avail(${socket})`));
    return ret;
}

// -----------------------
// Main Program
// -----------------------
(async () => {
    await Initialize("192.168.0.50", "192.168.0.1");
    const socket = 0;
    const port = 5000;

    while (true) {
        await sleep(1000); // 1000 ms

        await Listen(socket, port);
        console.log("Listening...");
        await Accept(socket);

        console.log("Socket connected!");
        await Send(socket, "Hello, I am DUELink\r\n");

        while (true) {
            await sleep(1); // 1 ms
            const read = await Available(socket);

            if (read !== 0) {
                const data = new Uint8Array(read);
                await Receive(socket, data);
                const msg = Buffer.from(data).toString('utf-8');
                if (msg) {
                    console.log(msg);
                }
            }
        }
    }
})();